facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+        match http_response.status() {
+            200 | 201 => {
+                if !last_chunk {
+                    return Err(anyhow!(
+                        "received HTTP 200 or 201 response with chunks remaining"
+                    ));
+                }
+            }
+            308 => {
+                if !http_response.has("Range") {
+                    return Err(anyhow!(
+                        "No range header in response from GCS: {:?}",
+                        http_response.into_string()
+                    ));
+                }
+                let range_header = http_response.header("Range").unwrap();
+                // The range header is like "bytes=111-222", and represents a
+                // range in the overall object, not the current chunk.
+                let range = range_header
+                    .strip_prefix("bytes=")
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let mut values = range.splitn(2, '-');
+                let start = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+                let end = values
+                    .next()
+                    .context(format!("unexpected Range header {}", range_header))?
+                    .parse::<usize>()
+                    .context(format!("unexpected Range header {}", range_header))?;
+
+                if start != self.object_upload_position {
+                    return Err(anyhow!("unexpected Range header {}", range_header));
+                }
+
+                self.object_upload_position = end + 1;
+
+                // If we have a little content left over, we can't just make
+                // another request, because if there's too little of it, Google
+                // will reject it. Instead, put the portion of the chunk that we
+                // didn't manage to upload back into self.buffer so it can be
+                // handled by a subsequent call to upload_chunk.
+                self.buffer
+                    .extend_from_slice(&chunk[self.object_upload_position - start..]);
+            }
+            _ => {
+                return Err(anyhow!(
+                    "failed to upload part to GCS: {} synthetic: {}\n{:?}",
+                    http_response.status(),
+                    http_response.synthetic(),
+                    http_response.into_string()
+                ))
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Write for StreamingTransferWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        // Write into memory buffer, and upload to GCS if we have accumulated
+        // enough content
+        self.buffer.extend_from_slice(buf);
+        if self.buffer.len() >= self.minimum_upload_chunk_size {
```

Comment:
Should be `while`, I think.

Reply:
Nope! Consider: upload size is 8 MiB, and you have 8 MiB + 100 kB in the buffer. You try to upload the whole buffer, but Google tells you they only accepted 8 MiB of it. If you then try to upload the remaining 100 kB, Google will reject it, because they don't allow chunks less than 256 kB unless it's the last chunk. So we intentionally don't drain the buffer here and only do that in `complete_upload`, where we are able to construct a `Content-Range` header that specifies the total size of the object, in which case Google lets us upload a chunk of less than 256 kilobytes.

Reply:
I don't understand. Are you saying the buffer grows indefinitely because it's never shrunk until `complete_upload`? Because this reads to me as if it is replacing buffer with a new zero-length Vec: https://github.com/abetterinternet/prio-server/pull/88/files#diff-465ffcef811e76d63ff9b6f105387456ffff0155b61560ead82aa6d32ef9d7e2R349-R351.

> Google tells you they only accepted 8 MiB of it. If you then try to upload the remaining 100 kB, Google will reject it,

In this case, you would drop out of the `while` loop because buffer.len() (i.e. 100kB) is not greater than minimum_upload_chunk_size.

Imagine the case where you're getting `.write()` calls with 100MB each, but Google is only accepting 50MB per `PUT`. You'd fall behind.

Reply:
Oh, of course, you're right. I changed it to while.